### PR TITLE
feat(trends): Allow histogram breakdowns for numeric values

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

The way histogram binning works is:

1. get_breakdown_values returns an array of values that form buckets for binning. There's no limiting on the number of breakdown values, only the number of buckets.
2. Assumes it's working with numbers. Everything else returns NULLs.
3. If the breakdown values returned were [10, 20, 30, 40], this means we have buckets: [10, 20), [20, 30), [30, infinity)
4. The buckets are chosen by quantiles
5. The final bucket is open ended above, and all values are truncated below to 2 decimal places, which ensures we never miss a value. The first and last quantile are always 0, and 1, or in other words min(all values), and max(all values), which ensures no values are missed.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/queries/trends/sql.py [hunk 1]
```diff
@@ -226,7 +241,20 @@
 """
 
 
-SESSION_BREAKDOWN_AGGREGATE_QUERY_SQL = """
+BREAKDOWN_HISTOGRAM_AGGREGATE_QUERY_SQL = """
+SELECT histogram(%(bin_count)s)({aggregate_operation}) AS histogram
```
Comment (neilkakkar) [inline posthog/queries/trends/sql.py:245]: delete

File: posthog/queries/trends/sql.py [hunk 2]
```diff
@@ -253,6 +281,11 @@
   {actions_query}
 """
 
+BREAKDOWN_HISTOGRAM_PROP_JOIN_SQL = """
```
Comment (neilkakkar) [inline posthog/queries/trends/sql.py:284]: can combine into above without too much pain?

File: posthog/queries/breakdown_props.py [hunk 3]
```diff
@@ -160,12 +178,20 @@ def _to_value_expression(
...
     direct_on_events: bool = False,
+    cast_as_float: bool = False,
 ) -> str:
-    if breakdown_type == "person":
-        return get_single_or_multi_property_string_expr(
+    if breakdown_type == "session":
+        if breakdown == "$session_duration":
+            # Return the session duration expression right away because it's already an number,
+            # so it doesn't need casting for the histogram case (like the other properties)
+            return f"{SessionQuery.SESSION_TABLE_ALIAS}.session_duration as value"
```
Comment (neilkakkar) [inline posthog/queries/breakdown_props.py:187]: is there a reason why we wouldn't want to cast this to float64 or null? Would be nice to not special case this and keep things uniform

File: posthog/queries/trends/breakdown.py [hunk 4]
```diff
@@ -178,7 +180,7 @@ def get_query(self) -> Tuple[str, Dict, Callable]:
             if self.entity.math_property == "$session_duration":
                 # TODO: When we add more person/group properties to math_property,
                 # generalise this query to work for everything, not just sessions.
-                content_sql = SESSION_BREAKDOWN_AGGREGATE_QUERY_SQL.format(
+                content_sql = SESSION_MATH_BREAKDOWN_AGGREGATE_QUERY_SQL.format(
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/queries/trends/breakdown.py:183]: Never mind, not the case, missed that there exists a `session_join` in the latter! Good change 🙌

File: posthog/queries/breakdown_props.py (hunk 3 above)
Comment (rcmarron) [reply posthog/queries/breakdown_props.py:187]: I thought our property extractions stuff always returns a string. Looking more now

**Review by rcmarron — APPROVED** [review]
Looks good to me. Noting two things for follow on PRs:
* Figure out how we're gonna do rounding
* Make sure we support properties that are sent up as numbers (I think this is covered by our tests)

**Comment by neilkakkar** [comment]
I think I've handled both those things?

For rounding, look here: https://github.com/PostHog/posthog/pull/10606/files#diff-853552bb238513734098fcee5e96bcbf3766d9c7fcd1ffe617648d652b462dbaR419 - we're doing `floor` upto 2 decimal places, anything further than that seems overkill? And the final bucket is left unbounded, (see point 5 in description).

and relevant test: https://github.com/PostHog/posthog/pull/10606/files#diff-d5712f2ef79f93e09640112635a790b74c551c4664e185a6fb43dbaf88e1c8fcR54 - it's 180.5 seconds.

---

And for the toFloat64 thing, just pushed 🙈 : https://github.com/PostHog/posthog/pull/10606/commits/d3c63eb61b11edcdb503454eaa1baedb58534239 - `toString` happily accepts strings lol

File: posthog/queries/breakdown_props.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/queries/breakdown_props.py]: Ah! Probably. So maybe this matters only for calculated properties, a.k.a. just duration.

I made them uniform anyway

**Pull request merged**